pyomo/core/base/block.py

```diff
@@ -703,14 +710,99 @@ def __delattr__(self, name):
+            """\
+            Block components do not support assignment or set_value().
+            Use the transfer_attributes_from() method to transfer the
+            components and public attributes from one block to another:
+                model.b[1].transfer_attributes_from(other_block)
+            """))
+
+    def transfer_attributes_from(self, src, guarantee_components=()):
+        """Set (override) the value of this Component Data
+
+        This removes all components assigned to this block and then
+        moves over all components and all non-private attributes from
+        `src`.  Because this component is not slotized, we cannot
+        distinguish between instance attributes declared by `__init__()`
+        and non-components assigned by the user.  Therefore, we will not
+        remove *any* attributes and only copy over attributes that
+        either are not already present here or are not private (do not
+        begin with a "_").
+
+        Derived blocks may wish to ensure that certain components are
+        always present on this block (notably the `indicator_var` `Var`
+        on `Disjunct`).  Derived classes may wrap this method and
+        provide a `guarantee_components` set.  Components whose local
+        name appears in the `guarantee_components` set will only be
+        removed from this Block if `src` contains either a component or
+        attribute with the same local name.  This will "guarantee" that
+        this object will still have the required attributes after
+        set_value()
+
+        Parameters
+        ----------
+        src: _BlockData or dict
+            The Block or mapping that contains the new attributes to
+            assign to this block.
+
+        guarantee_components: sequence or set
+            components on this block whose local name appears in
+            guarantee_components will not be automatically removed
+            unless there is a component or attribute in `src` with the
+            same name, thereby guaranteeing that the component will be
+            present in the block.
+        """
+        if isinstance(src, _BlockData):
+            # There is a special case where assigning a parent block to
+            # this block creates a circular hierarchy
+            if src is self:
+                return
+            p_block = self.parent_block()
+            while p_block is not None:
+                if p_block is src:
+                    raise ValueError(
+                        "_BlockData.transfer_attributes_from(): Cannot set a "
+                        "sub-block (%s) to a parent block (%s): creates a "
+                        "circular hierarchy" % (self, src))
+                p_block = p_block.parent_block()
+            # record the components and the non-component objects added
+            # to the block
+            src_comp_map = src.component_map()
+            src_raw_dict = src.__dict__
+        elif src is None:
```

Comment:
No.  You still need to go through and remove any components on this block that are not in the guarantee_components set (lines 783-787), effectively "clearing" this block.

Reply:
That seems like unexpected behavior to me, so the following:

```
b = Block()
b.x = Var()
b.y = Var()

b.transfer_attributes_from(None, guarantee_components={'x'})
```

would remove `y` from `b`? I guess I wouldn't think to use the `transfer_attributes_from` method to "clear" a block like that. I follow how it's an easy piece of functionality to include here to avoid duplicating code in a separate method but it strikes me as a loophole in the implementation instead of an intentional feature that would be obvious to users.

Reply:
The current behavior stems from this originally being implemented within `set_value()`.  You are arguing that we should move some of this functionality into `_BlockData.clear()`, and a user would just call `b.clear()`?

I think one of the challenges for b.clear() is to know *what* can be cleared.  The more I think about it, the more I think we should be able to leverage the `_Block_reserved_words` attribute.  That would - in theory - allow us to remove `guarantee_components` entirely.

Reply:
Yes, that is what I'm advocating, an explicit `b.clear()` call (or maybe `b.del_all_components()` to match our `del_component` name?). 

I'm not currently relying on this feature so if this implementation makes sense to you and @emma58, I don't want to hold up a PR that is blocking her progress. I just find it a little confusing from an outside perspective. I can open a follow-up issue to discuss this if you prefer.